oracle/oracle.go

```diff
@@ -0,0 +1,377 @@
+	if err != nil {
+		return nil, err
+	}
+	oracleOps.containerEngine, err = containerengine.NewContainerEngineClientWithConfigurationProvider(configProvider)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: wrap around exponentialBackoffOps
+	return oracleOps, nil
+}
+
+func getInfoFromEnv(oracleOps *oracleOps) error {
+	var err error
+	oracleOps.instance, err = cloudops.GetEnvValueStrict(envInstanceID)
+	if err != nil {
+		return err
+	}
+
+	oracleOps.region, err = cloudops.GetEnvValueStrict(envRegion)
+	if err != nil {
+		return err
+	}
+
+	oracleOps.availabilityDomain, err = cloudops.GetEnvValueStrict(envAvailabilityDomain)
+	if err != nil {
+		return err
+	}
+
+	oracleOps.compartmentID, err = cloudops.GetEnvValueStrict(envCompartmentID)
+	if err != nil {
+		return err
+	}
+
+	oracleOps.tenancyID, err = cloudops.GetEnvValueStrict(envTenancyID)
+	if err != nil {
+		return err
+	}
+
+	oracleOps.poolID, err = cloudops.GetEnvValueStrict(envPoolID)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func getRequest(endpoint string, headers map[string]string) (map[string]interface{}, int, error) {
+	metadata := make(map[string]interface{})
+	client := &http.Client{}
+	req, err := http.NewRequest("GET", endpoint, nil)
+	if err != nil {
+		return metadata, 0, err
+	}
+
+	for headerKey, headerValue := range headers {
+		req.Header.Add(headerKey, headerValue)
+	}
+	q := req.URL.Query()
+	req.URL.RawQuery = q.Encode()
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return metadata, resp.StatusCode,
+			fmt.Errorf("metadata lookup from [%s] endpoint failed with error:[%v]", endpoint, err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return metadata, resp.StatusCode, nil
+	}
+	if resp.Body != nil {
+		defer resp.Body.Close()
+		respBody, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return metadata, resp.StatusCode,
+				fmt.Errorf("error while reading Oracle metadata response: [%v]", err)
+		}
+		if len(respBody) == 0 {
+			return metadata, resp.StatusCode,
+				fmt.Errorf("error querying Oracle metadata: Empty response")
+		}
+
+		err = json.Unmarshal(respBody, &metadata)
+		if err != nil {
+			return metadata, resp.StatusCode,
+				fmt.Errorf("error parsing Oracle metadata: %v", err)
+		}
+	}
+	return metadata, resp.StatusCode, nil
+}
+
+// GetMetadata returns metadata from IMDS
+func GetMetadata() (map[string]interface{}, error) {
+	httpHeaders := map[string]string{}
+	httpHeaders["Authorization"] = "Bearer Oracle"
+	var httpStatusCode int
+	var err error
+	var metadata map[string]interface{}
+	metadata, httpStatusCode, err = getRequest(v2MetadataAPIEndpoint, httpHeaders)
+	if err != nil {
+		return nil, err
+	}
+
+	if httpStatusCode != http.StatusOK {
+		// Authorization header is safely ignored by v1 endpoint
+		logrus.Warnf("Trying %s endpoint as got %d http response from %s\n",
+			v2MetadataAPIEndpoint, httpStatusCode, v1MetadataAPIEndpoint)
+		metadata, httpStatusCode, err = getRequest(v1MetadataAPIEndpoint, httpHeaders)
```

Comment:
I understand that it ignores the header. I don't get why we want to pass it though. Can't we just pass a blank map here and remove line 179?

Reply:
Done